polytope_server/common/metric_store/dynamodb_metric_store.py

```diff
@@ -0,0 +1,207 @@
+#
+# Copyright 2024 European Centre for Medium-Range Weather Forecasts (ECMWF)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation nor
+# does it submit to any jurisdiction.
+#
+
+import json
+import logging
+import operator
+from decimal import Decimal
+from enum import Enum
+from functools import reduce
+
+import boto3
+import botocore
+from boto3.dynamodb.conditions import Attr, Key
+import botocore.exceptions
+
+from ..metric import (
+    CacheInfo,
+    Metric,
+    MetricType,
+    QueueInfo,
+    RequestStatusChange,
+    StorageInfo,
+    WorkerInfo,
+    WorkerStatusChange,
+)
+from . import MetricStore
+
+logger = logging.getLogger(__name__)
+
+
+METRIC_TYPE_CLASS_MAP = {
+    MetricType.WORKER_STATUS_CHANGE: WorkerStatusChange,
+    MetricType.WORKER_INFO: WorkerInfo,
+    MetricType.REQUEST_STATUS_CHANGE: RequestStatusChange,
+    MetricType.STORAGE_INFO: StorageInfo,
+    MetricType.CACHE_INFO: CacheInfo,
+    MetricType.QUEUE_INFO: QueueInfo,
+}
+
+
+def _iter_items(fn, **params):
+    while True:
+        response = fn(**params)
+        for item in response["Items"]:
+            yield item
+        if "LastEvaluatedKey" not in response:
+            break
+        params["ExclusiveStartKey"] = response["LastEvaluatedKey"]
+
+
+def _make_query(**kwargs):
+    return {
+        key: value.value if isinstance(value, Enum) else value for key, value in kwargs.items() if value is not None
+    }
+
+def _visit(obj, fn):
+    if isinstance(obj, dict):
+        return {key: _visit(value, fn) for key, value in obj.items()}
+    if isinstance(obj, list):
+        return [_visit(value, fn) for value in obj]
+    return fn(obj)
+
+
+def _convert_numbers(obj, reverse=False):
+    def fn(item):
+        if not reverse and isinstance(item, float):
+            return Decimal(item)
+        elif reverse and isinstance(item, Decimal):
+            return float(item)
+        return item
+    return _visit(obj, fn)
+
+
+def _load(item):
+    metric_type = Metric.deserialize_slot("type", item["type"])
```

Comment:
It could be a case where the item is null or the key doesn't exist and could rise an exception.

Do you think may be worth to check before hand?

e.x:

```
if "type" not in item:
    raise ValueError("Metric type not found in item")
```

Reply:
The current behaviour is to raise a KeyError which I feel is more relevant in this case. The type attribute is part of the base class and should be present for all items in the database but I agree that a malformed metric should not cause the server to crash.